File: dex.rb
require 'find'
require 'fileutils'

# Define the source and destination directories
source_directory = '.'
destination_file = 'DEX.md'

# Initialize an array to store the formatted headers
formatted_headers = []

# Regular expression to match the datetime format
datetime_regex = /\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}/

# Iterate through the files in the source directory
Find.find(source_directory) do |file_path|
  next unless File.file?(file_path) && file_path =~ /\.md$/

  # Read the file content
  content = File.read(file_path)

  # Extract the first header (assuming it's the only one) using regex
  header = content.match(/^# (.+)$/)&.captures&.first

  # Check if a valid header was found and if it contains a datetime
  if header && header.match(datetime_regex)
    # Remove any leading/trailing whitespace, colons, and question marks,
    # and convert to lowercase
    formatted_header = header.strip.downcase.gsub(/[?:]/, '')

    # Replace spaces with hyphens
    formatted_header.gsub!(' ', '-')

    # Append the formatted header to the array
    formatted_headers << "[[#{formatted_header}]]"
  end
end

# Write the formatted headers to the destination file
File.open(destination_file, 'w') do |file|
  file.puts("# DEX\n\n")
  formatted_headers.each { |line| file.puts(line) }
end

puts "Formatted headers written to #{destination_file}"

File: dex.js
const fs = require('fs');
const path = require('path');

// Define the source and destination directories
const sourceDirectory = './notes';
const destinationFile = 'DEX.md';

// Initialize an array to store the formatted headers
const formattedHeaders = [];

// Regular expression to match the datetime format
const datetimeRegex = /\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}/;

// Function to process a Markdown file and extract its header
function processFile(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  const matches = content.match(/^# (.+)$/m);

  if (matches && matches.length > 1) {
    const header = matches[1];

    if (header.match(datetimeRegex)) {
      const formattedHeader = header.trim().toLowerCase().replace(/[?:]/g, '').replace(/ /g, '-');
      formattedHeaders.push(`[[${formattedHeader}]]`);
    }
  }
}

// Recursive function to traverse the directory and process files
function traverseDirectory(directoryPath) {
  const files = fs.readdirSync(directoryPath);

  for (const file of files) {
    const filePath = path.join(directoryPath, file);
    const stat = fs.statSync(filePath);

    if (stat.isDirectory()) {
      traverseDirectory(filePath);
    } else if (stat.isFile() && file.endsWith('.md')) {
      processFile(filePath);
    }
  }
}

// Start the directory traversal
traverseDirectory(sourceDirectory);

// Add the DEX header
formattedHeaders.unshift(`# DEX`);

// Write the formatted headers to the destination file
fs.writeFileSync(destinationFile, formattedHeaders.join('\n'), 'utf8');

console.log(`Formatted headers written to ${destinationFile}`);
